src/java.xml.crypto/share/classes/javax/xml/crypto/dsig/spec/RSAPSSParameterSpec.java

```diff
@@ -0,0 +1,145 @@
+import java.security.spec.MGF1ParameterSpec;
+import java.security.spec.PSSParameterSpec;
+import java.util.Objects;
+
+/**
+ * Parameters for the <a href="http://www.w3.org/2007/05/xmldsig-more#rsa-pss">
+ * XML Signature RSASSA-PSS Algorithm</a>. The parameters are expressed as a
+ * {@link PSSParameterSpec} object encapsulated.
+ * <p>
+ * The XML Schema Definition is defined as:
+ * <pre><code>
+ *    &lt;xs:element name="RSAPSSParams" type="pss:RSAPSSParamsType"&gt;
+ *        &lt;xs:annotation&gt;
+ *            &lt;xs:documentation&gt;
+ *    Top level element that can be used in xs:any namespace="#other"
+ *    wildcard of ds:SignatureMethod content.
+ *            &lt;/xs:documentation&gt;
+ *        &lt;/xs:annotation&gt;
+ *    &lt;/xs:element&gt;
+ *    &lt;xs:complexType name="RSAPSSParamsType"&gt;
+ *        &lt;xs:sequence&gt;
+ *            &lt;xs:element ref="ds:DigestMethod" minOccurs="0"/&gt;
+ *            &lt;xs:element name="MaskGenerationFunction"
+ *               type="pss:MaskGenerationFunctionType" minOccurs="0"/&gt;
+ *            &lt;xs:element name="SaltLength" type="xs:int"
+ *               minOccurs="0"/&gt;
+ *            &lt;xs:element name="TrailerField" type="xs:int"
+ *               minOccurs="0"/&gt;
+ *        &lt;/xs:sequence&gt;
+ *    &lt;/xs:complexType&gt;
+ *    &lt;xs:complexType name="MaskGenerationFunctionType"&gt;
+ *        &lt;xs:sequence&gt;
+ *            &lt;xs:element ref="ds:DigestMethod" minOccurs="0"/&gt;
+ *        &lt;/xs:sequence&gt;
+ *        &lt;xs:attribute name="Algorithm" type="xs:anyURI"
+ *           default="http://www.w3.org/2007/05/xmldsig-more#MGF1"/&gt;
+ *    &lt;/xs:complexType&gt;
+ * </code></pre>
+ *
+ * @implSpec
+ * If the parameter is not specified when using the RSASSA-PSS signature
+ * algorithm, the default parameter is used, which uses SHA-256 as the
+ * {@code DigestMethod}, MGF1 with SHA-256 as the
+ * {@code MaskGenerationFunction}, 32 as {@code SaltLength}, and 1 as
+ * {@code TrailerField}. This is equivalent to the parameter-less signature
+ * method as defined by http://www.w3.org/2007/05/xmldsig-more#sha256-rsa-MGF1.
+ *
+ * @since 17
+ * @see SignatureMethod
+ * @see <a href="https://www.ietf.org/rfc/rfc6931.txt">RFC 6931</a>
+ */
+public final class RSAPSSParameterSpec implements SignatureMethodParameterSpec {
+
+    private final PSSParameterSpec spec;
+
+    /**
+     * Creates a new {@code RSAPSSParameterSpec} object with the specified
+     * {@link PSSParameterSpec} object encapsulated.
+     *
+     * @param spec the input {@code PSSParameterSpec} to be encapsulated
+     */
+    public RSAPSSParameterSpec(PSSParameterSpec spec) {
+        this.spec = Objects.requireNonNull(spec);
+    }
+
+    /**
+     * Returns the encapsulated {@code PSSParameterSpec} object.
+     *
+     * @return the encapsulated {@code PSSParameterSpec} object
+     */
+    public PSSParameterSpec getPSSParameterSpec() {
```

Comment:
If an XML Signature contained an RSAPSSParams with no DigestMethod, would this return a PSSParameterSpec with the defaults as specified in the @implSpec?

Reply:
Yes. The `DOMRSAPSSSignatureMethod::unmarshalParams` method will fill the blanks.

Reply:
I wonder if the @implSpec is clear enough that this will be returned. I might suggest adding a similar @implSpec in this method that basically states what you said above.

Reply:
I'm not sure if it's appropriate to specify the default value in this method. As long as there is an `RSAPSSParameterSpec` object, there must be a non-null `PSSParameterSpec` inside and it is the one that was used to construct this object.

I am thinking if we can append the following to the `@implNote` in the class spec:
```
 * One can obtain this default value using the following expression:
 * <pre><code>
 * XMLSignatureFactory.getInstance()
 *      .newSignatureMethod(SignatureMethod.RSA_PSS, null)
 *      .getParameterSpec()
 * </code></pre>
```

but this makes it more like an `@implNote` instead of an `@implSpec`.

Reply:
I added the new lines as `@implNote` and kept the old `@implSpec` there (since it's still a requirement for implementations). New commit pushed. CSR updated as well.

Reply:
Ok, I understand now. I think `@implSpec` (and probably the `@implNote`) are in the wrong class. `@implSpec`  means the implementation of this class. But this class is final and does not contain that logic. The logic of specifying/returning the defaults is in the JDK (XMLDSig provider) implementation of `SignatureMethod`. So I think it belongs there. In this class, you could add a sentence/link to `SignatureMethod`, something like "See SignatureMethod for how default values are determined when the parameters are not specified." 

Also, I think the `@implSpec` needs to be more specific, and also cover the cases where some, but not all of the parameters are specified and defaults are then used. For this, you will need to be more general, as the default saly length is based on what hash algorithm is being used. 

As for `@implNote`, this probably could use more discussion, but it might be better to make this part of the specification. If some implementations can return null, and others return defaults, it complicates the application's logic. The RFC has clearly specified what the defaults should be, so maybe the easiest thing to do is to make all implementations comply by also making it part of the API contract, and hiding the XML details as to whether the parameter was actually specified or not, which should not matter to applications.

Reply:
You are right that the overridable methods are elsewhere (`XMLSignatureFactory::newSignatureMethod` and `SignatureMethod::getParameterSpec`), but I still feel it a little strange to move the default parameter of one particular algorithm to the general interface `SignatureMethod` (this is similar to talking about EC curve names in `KeyPairGenerator`). It seems more natural to talk about this inside a class which is specific to the RSASSA-PSS algorithm and `RSAPSSParameterSpec` is the only public API we can find. We can add something like "specify null if you want a default spec" to `XMLSignatureFactory::newSignatureMethod` but it does not have enough space to describe "how default values are determined" for each algorithm.

I understand `@implSpec` is for implementations of a class or a method, but does not mean the words must be put inside that exact class and method? Maybe not necessarily.

As for making `@implSpec` more specific, at signing time, we can only either provide a `RSAPSSParameterSpec` or not one, so there is only one default value which is SHA256_RSA_MGF1. On the other hand, at validation time we might be parsing a partial-filled `RSAPSSParams` node and that's where default salt and default trailer field show up.

Also about the return value of `SignatureMethod::getParameterSpec`, are you suggesting that for RSASSA-PSS it must be non null? This can be specified somewhere but we need to find a place. For the existing `HMACParameterSpec`, the method can return null if none is set? Do we treat that as no spec at all (but for RSASSA-PSS there is always one)?

My current opinion is that we still document all these in `RSAPSSParameterSpec` but be very clear that this part is for `XMLSignatureFactory::newSignatureMethod` and that part is for `SignatureMethod::getParameterSpec`, etc, etc.

Reply:
I think it is worth asking the TCK team about this. After further thought, I am not sure `implSpec` is correct because the implementation is not in the classes, it is in the provider. I now think it needs to be part of the API contract, so that all implementations are compliant. `SignatureMethod` already defines the RSA_PSS algorithm constant, so it doesn't seem so out-of-place to put the specification there, something like:

```
/**
  * The <a href="http://www.w3.org/2007/05/xmldsig-more#rsa-pss">
  * RSASSA-PSS</a> signature method algorithm URI.
  *
  * If the parameter is not specified when calling `XMLSignatureFactory.newSignatureMethod` with 
  * RSA_PSS as the signature algorithm, the default parameter is used, which uses SHA-256 as the
  * {@code DigestMethod}, MGF1 with SHA-256 as the
  * {@code MaskGenerationFunction}, 32 as {@code SaltLength}, and 1 as
  * {@code TrailerField}. This is equivalent to the parameter-less signature
  * method {@link SignatureMethod#SHA256_RSA_MGF1 SHA256_RSA_MGF1} as defined
  * in <a href="https://tools.ietf.org/html/rfc6931#section-2.3.10">RFC 6931. This default
  * parameter is returned by the `getParameterSpec` method.
  *
  * @since 17
  */
 String RSA_PSS = "http://www.w3.org/2007/05/xmldsig-more#rsa-pss";
```
Forget my comment about making `implSpec` more specific, I think that is now implied by RFC 3161. I also think the above specification would take care of my returning null comment, as all implementations would need to comply. I just didn't want applications to have to have code that checks for null and then don't know whether that means it is the default parameters or something else since it was implementation specific.

Reply:
Great, I forgot about this string. This *is* a proper place to put the text.